classes/external/get_join_url.php

```diff
@@ -0,0 +1,121 @@
+use restricted_context_exception;
+
+defined('MOODLE_INTERNAL') || die();
+
+global $CFG;
+require_once($CFG->libdir . '/externallib.php');
+
+/**
+ * External service to create the meeting (if needed), check user limit, and return the join URL when we can join.
+ *
+ * This is mainly used by the mobile application.
+ *
+ * @package   mod_bigbluebuttonbn
+ * @category  external
+ * @copyright 2018 onwards, Blindside Networks Inc
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class get_join_url extends external_api {
+    /**
+     * Returns description of method parameters
+     *
+     * @return external_function_parameters
+     */
+    public static function execute_parameters(): external_function_parameters {
+        return new external_function_parameters([
+            'cmid' => new external_value(PARAM_INT, 'course module id', VALUE_REQUIRED),
+            'groupid' => new external_value(PARAM_INT, 'bigbluebuttonbn group id', VALUE_DEFAULT, 0),
+        ]);
+    }
+
+    /**
+     * Updates a recording
+     *
+     * @param int $cmid the bigbluebuttonbn course module id
+     * @param null|int $groupid
+     * @return array (empty array for now)
+     * @throws \restricted_context_exception
+     */
+    public static function execute(
+        int $cmid,
+        ?int $groupid = 0
+    ): array {
+        // Validate the cmid ID.
+        // TODO: we should maybe pass the bbbid + groupid instead of the cmid.
+        [
+            'cmid' => $cmid,
+            'groupid' => $groupid,
+        ] = self::validate_parameters(self::execute_parameters(), [
+            'cmid' => $cmid,
+            'groupid' => $groupid,
+        ]);
+        $result['warnings'] = [];
+
+        $instance = instance::get_from_cmid($cmid);
+        if (empty($instance)) {
+            throw new \moodle_exception('nosuchinstance', 'mod_bigbluebuttonbn', null,
+                ['entity' => get_string('module', 'course'), 'id' => $cmid]);
+        }
+        if (!groups_group_visible($groupid, $instance->get_course(), $instance->get_cm())) {
+            throw new restricted_context_exception();
+        }
+        $instance->set_group_id($groupid);
+
+        self::validate_context($instance->get_context());
+
+        $returnvalue =
+            meeting::create_and_join_meeting($instance);
+        if (!empty($returnvalue['url']) && empty($returnvalue['warningcode'])) {
+            $result['join_url'] = $returnvalue['url']; // We only return the joinURL if there is no warning or error.
+        }
+        if (!empty($returnvalue['warningcode'])) {
+            $result['warnings'][] = [
+                'item' => 'mod_bigbluebuttonbn',
+                'itemid' => $instance->get_instance_id(),
+                'warningcode' => $returnvalue['warningcode'],
+                'message' => get_string($returnvalue['warningcode'], 'mod_bigbluebuttonbn')
+            ];
+        }
+        if (!empty($returnvalue['errorcode'])) {
```

Comment:
As noted in the function definition, this should be a try/catch.

Reply:
Not totally sure I understand the comment, but I guess the main issue here is that instead of returning an error we should just raise an exception in the create_and_join_meeting. Is it what you meant ?

Reply:
Yes. We never use the URL if there is an error or a warning, so why have both in the array.